### PR TITLE
added setConnectionManager(bPooling) to SpotifyApi.Builder

### PR DESCRIPTION
**Pull request by gahisee**
if bPooling is true, PoolingHttpClientConnectionManager is used.
if not called, default is BasicHttpClientConnectionManager.
This addition allows for certain applications where multiple concurrent requests are made.
Otherwise, the following exception will occur:
java.lang.IllegalStateException: Connection 192.168.1.13:35776<->35.186.224.25:443 is still allocated
        at org.apache.hc.core5.util.Asserts.check(Asserts.java:50)

### CODE REVIEW

File: src/main/java/se/michaelthelin/spotify/SpotifyApi.java [hunk 1]
```diff
@@ -72,9 +72,9 @@ public class SpotifyApi {
   public static final String DEFAULT_HOST = "api.spotify.com";
 
   /**
-   * A HttpManager configured with default settings.
+   * A HttpManager Builder ready to build SpotifyHttpManager
    */
-  public static final IHttpManager DEFAULT_HTTP_MANAGER = new SpotifyHttpManager.Builder().build();
+  public static final SpotifyHttpManager.Builder shmb = new SpotifyHttpManager.Builder();
```
Comment (dargmuesli) [inline src/main/java/se/michaelthelin/spotify/SpotifyApi.java:77]: Please try to name variables without abbreviations.

File: src/main/java/se/michaelthelin/spotify/SpotifyHttpManager.java [hunk 2]
```diff
@@ -59,6 +61,9 @@ public SpotifyHttpManager(Builder builder) {
     this.connectionRequestTimeout = builder.connectionRequestTimeout;
     this.connectTimeout = builder.connectTimeout;
     this.socketTimeout = builder.socketTimeout;
+    this.usePooledConnectionManager = builder.usePooledConnectionManager;
+    PoolingHttpClientConnectionManager connectionManagerPooling = null;
+    BasicHttpClientConnectionManager connectionManagerBasic = null;
```
Comment (dargmuesli) [inline src/main/java/se/michaelthelin/spotify/SpotifyHttpManager.java:66]: You should be able to use `HttpClientConnectionManager` to collapse both variables into one.

File: src/main/java/se/michaelthelin/spotify/SpotifyApi.java [hunk 3]
```diff
@@ -1899,6 +1900,18 @@ public static class Builder {
     private String accessToken;
     private String refreshToken;
 
+    /**
+     * setConnectionManager.
```
Comment (dargmuesli) [inline src/main/java/se/michaelthelin/spotify/SpotifyApi.java:1904]: Please add proper documentation if any :pray:

File: src/main/java/se/michaelthelin/spotify/SpotifyApi.java [hunk 4]
```diff
@@ -1899,6 +1900,18 @@ public static class Builder {
...
+     *
+     * @return A {@link Builder}.
+     */
+    public Builder setConnectionManager(boolean bPooling) {
+      if (bPooling)
+        shmb.usePooledConnectionManager();
+      httpManager = shmb.build();
```
Comment (dargmuesli) [inline src/main/java/se/michaelthelin/spotify/SpotifyApi.java:1911]: Do we still need this code? `build` is also used in `public SpotifyApi build()` below.

File: src/main/java/se/michaelthelin/spotify/SpotifyHttpManager.java [hunk 5]
```diff
@@ -402,6 +421,11 @@ public Builder setSocketTimeout(Integer socketTimeout) {
       return this;
     }
 
+    public Builder usePooledConnectionManager() {
```
Comment (dargmuesli) [inline src/main/java/se/michaelthelin/spotify/SpotifyHttpManager.java:424]: Builder methods should be setter only for now.

File: spotify-web-api-java.jar
Comment (dargmuesli) [inline spotify-web-api-java.jar:1]: I think this file should not be committed into the source.

File: src/main/java/se/michaelthelin/spotify/SpotifyHttpManager.java [hunk 6]
```diff
@@ -81,9 +86,24 @@ public SpotifyHttpManager(Builder builder) {
...
-    BasicHttpClientConnectionManager connectionManager = new BasicHttpClientConnectionManager();
-    connectionManager.setConnectionConfig(connectionConfig);
-    RequestConfig requestConfig = RequestConfig
+
+      if (usePooledConnectionManager)
+      {
+        connectionManager = new PoolingHttpClientConnectionManager();
+        SpotifyApi.LOGGER.log(
+          Level.FINE,
+          "SpotifyHttpManager - using PoolingHttpClientConnectionManager.");
```
Comment (dargmuesli) [inline src/main/java/se/michaelthelin/spotify/SpotifyHttpManager.java:95]:
```suggestion
          "Using PoolingHttpClientConnectionManager");
```

Same below. The logger should add the class name as prefix already, I think.

**Comment by dargmuesli** [comment]
Please respond to the threads when making changes, indicating which threads can be resolved so I can verify and resolve :raised_hands:

File: spotify-web-api-java.jar
Comment (gahisee) [reply spotify-web-api-java.jar:1]: Agree. It has been removed.

File: src/main/java/se/michaelthelin/spotify/SpotifyHttpManager.java (hunk 6 above)
Comment (gahisee) [reply src/main/java/se/michaelthelin/spotify/SpotifyHttpManager.java:95]: OK.

File: spotify-web-api-java.jar
Comment (dargmuesli) [reply spotify-web-api-java.jar:1]: It appears you pushed to the `backward-compatibility` branch only for which we closed the PR. This PR needs the commit pushed to branch `poolingConnectionManager`.

Comment (gahisee) [reply spotify-web-api-java.jar:1]: I'm using GitKraken for this work and have gotten into a real mess working with 2 branches! Keep encountering "Detached HEAD" situation.  But I have managed to update the original branch and pushed it.  Appreciate your patience with my git struggles...

Comment (dargmuesli) [reply spotify-web-api-java.jar:1]: Thank you for being honest and transparent about being new to these PRs! :rocket:

**Review by ulpcan — APPROVED** [review]

**Comment by ulpcan** [comment]
I am getting this error too much, looking good. When do you plan to release?

File: src/main/java/se/michaelthelin/spotify/SpotifyHttpManager.java [hunk 7]
```diff
@@ -81,9 +86,22 @@ public SpotifyHttpManager(Builder builder) {
...
+
+    if (usePooledConnectionManager)
+    {
+      connectionManagerPooling = new PoolingHttpClientConnectionManager();
+      connectionManagerPooling.setMaxTotal(5);
+      connectionManagerPooling.setDefaultMaxPerRoute(2);
+      System.out.print("\n\n....... [SpotifyHttpManager] using PoolingConnectionManager....\n\n");
+    }
+    else
+    {
```
[the remark this answers is not in the document]
Comment (dargmuesli) [reply src/main/java/se/michaelthelin/spotify/SpotifyHttpManager.java]: I did a reformatting using IntelliJ IDEA

**Comment by dargmuesli** [comment]
@gahisee what do you think about just always using the pooling connection manager?

We could maybe safe on some complexity by just using that instead of the basic connection manager, or do you see any breaking changes?

**Comment by dargmuesli** [comment]
@ulpcan have you ever tried out the changes proposed in this PR in a project of yours or similar to confirm them to be working as is or have you just approved the changes without checking?

**Comment by gahisee** [comment]
@dargmuesli i always like having options. With one call, the user can opt for pooling manager;
otherwise, default is the basic one-connection manager as always. If we always use the pooling manager, presumably there would be a bigger memory impact though likely inconsequential.

**Comment by Selbi182** [comment]
Just wanna drop in and say I'm getting the same issue. It doesn't seem to affect the uptime of my app, but it does result in a lot of console spam:

grafik

**Pull request merged**